schemas: wakeup-source: Possibility for power states

scosu opened the pull request:
wakeup-source in its current form ignores power states. But with modern SoCs there are many powerstates that have a different set of devices enabled. Also there are more complex wakeup mechanisms available.

As an example TI's am62 SoC has a power state called 'Partial-IO' in which most of the SoC is powered off and only a small device logic is present for a few select devices that can still wakeup the SoC. But most devices that are usually able to wakeup the SoC are not capable to do it in Partial-IO.

In order to properly describe the wakeup-source property depending on the power state, this patch extends the definition to be either boolean or a list of power state names in which this device is capable of wakeup.

This is the Linux Kernel series that adds support for Partial-IO:
https://lore.kernel.org/all/20241219-topic-am62-partialio-v6-12-b4-v4-0-1cb8eabd407e@baylibre.com/

This is was my previous patch adding this binding before realizing the wakeup-source property is defined in this schema repository:
https://lore.kernel.org/all/20241219-topic-mcan-wakeup-source-v6-12-v6-1-1356c7f7cfda@baylibre.com/

krzk commented on dtschema/schemas/wakeup-source.yaml:
```diff
@@ -30,7 +32,9 @@ select: true
-    type: boolean
+    oneOf:
+      - type: boolean
+      - $ref: /schemas/types.yaml#/definitions/string-array
```
You need to constrain the values somewhere. Either here or in device schema (which I do not see you doing here https://lore.kernel.org/all/20241219-topic-am62-partialio-v6-12-b4-v4-2-1cb8eabd407e@baylibre.com/ ). If here, then the next question is - what is suspend? What sort of suspend? Shall we include all known/supported by Linux? s2idle? s2ram? or even ACPI states?

Maybe 'suspend' would mean all of them and if anyone more fine-grained then we add 'suspend-to-ram'?

Or maybe we should use something from Documentation/devicetree/bindings/cpu/idle-states.yaml ?

scosu replied:
I looked into the cpu/idle-states.yaml and Documentation/devicetree/bindings/power/domain-idle-state.yaml as well. I think these would be a flexible way to define the states without it being hardcoded strings. But they require entry/exit latency which does not make sense for Partial-IO as it is a power-off state. It can't be resumed, just wakeup and do a normal boot.

Maybe we could drop some of the required properties to be able to describe it as part of these idle states?
Or introduce poweroff-states that describe states that do not offer a resume path, just a boot? wakeup-source could then be a list of phandles to one of these nodes each?

krzk replied:
> Maybe we could drop some of the required properties to be able to describe it as part of these idle states?
Or introduce poweroff-states that describe states that do not offer a resume path, just a boot? wakeup-source could then be a list of phandles to one of these nodes each

I think domain-idle-states are specific to PM domains, so when I proposed to use their approach I thought to rather build on top or rework, not use directly. You do not need entry/exit latencies now, but we might need them in the future, so as you said - hard-coded strings are a bit limited.

I think you could go with entirely new bindings "system-idle-states" with as little properties as just name ("idle-state-name"). This can grow later. 

I do not see your patchset introducing DTS user at all (and this will a blocker also for dtschema - why do we want all this if there is no DTS user?), so tricky to say what you still need, but I guess you need to differentiate between suspend/wakeup states? If yes either by idle-state-name or by compatible.

scosu commented:
I updated these patches to add system-idle-state.

The updated series adding Partial-IO support uses system-idle-states and wakeup-source lists pointing to the states. The new version is here:
https://lore.kernel.org/all/20250306-topic-am62-partialio-v6-12-b4-v5-0-f9323d3744a2@baylibre.com/

krzk commented on dtschema/schemas/system-idle-states.yaml:
```diff
@@ -0,0 +1,53 @@
+
+    properties:
+      compatible:
+        const: "system-idle-state"
```
Drop quotes

krzk commented on dtschema/schemas/system-idle-states.yaml:
```diff
@@ -0,0 +1,53 @@
+        $ref: /schemas/types.yaml#/definitions/string
+        description:
+          Name of the system state.
+
```
If you do not have any properties here, are you expecting the name to define the type of the state? The idle-state-name in idle-states.yaml is just descriptive, like label. You want the value to be the ABI, right?

scosu replied:
Yes, this should be ABI. Should I change the description to make this clear?

robherring replied:
If it is an ABI, 'idle-state-name' should be required as well.

krzk commented on dtschema/schemas/wakeup-source.yaml:
```diff
@@ -69,4 +76,14 @@ examples:
+        reg = <0x00 0x4e08000 0x00 0x200>,
+              <0x00 0x4e00000 0x00 0x8000>;
+        reg-names = "m_can", "message_ram";
+        wakeup-source = "suspend", "poweroff";
```
This should be phandles.

krzk commented:
I think I don't know how to use Github or you pushed the same stuff. I don't see any improvements in latest commit mentioned above:
https://github.com/devicetree-org/dt-schema/pull/150/commits/807444d5169a59e22a50c0fa289a1cda1d18830e

scosu commented:
> I think I don't know how to use Github or you pushed the same stuff. I don't see any improvements in latest commit mentioned above: [807444d](https://github.com/devicetree-org/dt-schema/commit/807444d5169a59e22a50c0fa289a1cda1d18830e)

I now removed the quotes and fixed the missing phandles. But I don't know how to fix your comment regarding `idle-state-name`. I tried to respond to your comment but it may have not made it to you:

> Yes, this should be ABI. Should I change the description to make this clear?

scosu commented:
Thank you! I think I fixed all your comments.

robherring commented on dtschema/schemas/system-idle-states.yaml:
```diff
@@ -0,0 +1,54 @@
+      };
+
+      system-off {
+        compatible = "system-idle-state";
```
This will now fail validation as idle-state-name is missing

scosu replied:
Indeed, sorry, I missed that. Fixed

krzk commented on dtschema/schemas/wakeup-source.yaml:
```diff
@@ -69,4 +76,14 @@ examples:
+              <0x00 0x4e00000 0x00 0x8000>;
+        reg-names = "m_can", "message_ram";
+        wakeup-source = <&system_partial_io>, <&system_deep_sleep>;
+        status = "disabled";
```
Examples cannot be disabled. What is the reason behind disabling it?

krzk commented on dtschema/schemas/system-idle-states.yaml:
```diff
@@ -0,0 +1,55 @@
+      idle-state-name:
+        $ref: /schemas/types.yaml#/definitions/string
+        description:
+          Name of the system state.
```
Expanding my previous review. Do the actual values here matter? Can anyone use "pony", "suspend" or "suspend2ram" and be fine? I am asking (again) because I had impression you expect very specific values and if any other value is used, this will break.

That's different than existing meaning of idle-state-name, because before it was just like label, so purely informational.

scosu replied:
The actual value matters as it's effectively an ABI, it should be the same as described in the datasheet. In my implementation one driver uses that value and checks if it has a specific value (in that case it checks for 'Partial-IO'). I am also describing other states for completeness which are not used by any driver yet. https://lore.kernel.org/lkml/20250421-topic-am62-dt-partialio-v6-15-v1-2-6ced30aafddb@baylibre.com/

I could make this a list of possible values to restrict the possible strings here. But the names are probably mostly SoC family specific so they may have a meaning for one SoC family but would be invalid for another. Would that work for you?

krzk reviewed (approved):
Reviewed-by: Krzysztof Kozlowski <krzk@kernel.org>
